Hnsw format testRecall failing

### Description

After a6a96cde1c6 Bugfix/fix hnsw search termination check (#14215) HNSW format recall tests started failing. Need to investigate.

```
TestLucene94HnswVectorsFormat > testRecall FAILED
    java.lang.AssertionError: Average recall for DOT_PRODUCT should be at least 40.0 / 80, got 37
        at __randomizedtesting.SeedInfo.seed([F3F55E436AC90BCA:3E42FDA3F038909D]:0)
        at junit@4.13.1/org.junit.Assert.fail(Assert.java:89)
        at junit@4.13.1/org.junit.Assert.assertTrue(Assert.java:42)
        at org.apache.lucene.test_framework@11.0.0-SNAPSHOT/org.apache.lucene.tests.index.BaseKnnVectorsFormatTestCase.assertRecall(BaseKnnVectorsFormatTestCase.java:2009)
        at org.apache.lucene.test_framework@11.0.0-SNAPSHOT/org.apache.lucene.tests.index.BaseKnnVectorsFormatTestCase.testRecall(BaseKnnVectorsFormatTestCase.java:1940)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
        at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
        at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
        at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
```

This makes me think that we have many duplicate vectors in our test graph.

### Gradle command to reproduce

```
./gradlew :lucene:backward-codecs:test --tests "org.apache.lucene.backward_codecs.lucene94.TestLucene94HnswVectorsFormat.testRecall" -Ptests.jvms=12 -Ptests.jvmargs= -Ptests.seed=F3F55E436AC90BCA -Ptests.useSecurityManager=true -Ptests.gui=true -Ptests.file.encoding=UTF-8 -Ptests.vectorsize=128 -Ptests.forceintegervectors=true
```

--- a/lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java
+++ b/lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java
@@ -2047,10 +2047,9 @@ public abstract class BaseKnnVectorsFormatTestCase extends BaseIndexFileFormatTe
           if (line.isEmpty()) {
             continue;
           }
-          if (seen.contains(line)) {
+          if (seen.add(line) == false) {
             continue;
           }
-          seen.add(line);
           ++lineNo;
           Document doc = new Document();
           doc.add(

--- a/lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java
+++ b/lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java
@@ -1952,7 +1952,7 @@ public abstract class BaseKnnVectorsFormatTestCase extends BaseIndexFileFormatTe
       // indexed 421 lines from LICENSE.txt
       // indexed 157 lines from NOTICE.txt
       int topK = 10;
-      int numQueries = 578;
+      int numQueries = 526;
       String[] testQueries = {
         "Apache Lucene",
         "Apache License",
@@ -2036,6 +2036,7 @@ public abstract class BaseKnnVectorsFormatTestCase extends BaseIndexFileFormatTe
     Directory indexStore = newDirectory(random());
     IndexWriter writer = new IndexWriter(indexStore, newIndexWriterConfig());
     float[] scratch = new float[dimension];
+    Set<String> seen = new HashSet<>(578);
     for (String file : List.of("LICENSE.txt", "NOTICE.txt")) {
       try (InputStream in = BaseKnnVectorsFormatTestCase.class.getResourceAsStream(file);
           BufferedReader reader = new BufferedReader(new InputStreamReader(in, UTF_8))) {
@@ -2046,6 +2047,10 @@ public abstract class BaseKnnVectorsFormatTestCase extends BaseIndexFileFormatTe
           if (line.isEmpty()) {
             continue;
           }
+          if (seen.contains(line)) {
+            continue;
+          }
+          seen.add(line);
           ++lineNo;
           Document doc = new Document();
           doc.add(

